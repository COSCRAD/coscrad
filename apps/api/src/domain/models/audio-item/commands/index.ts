export * from './constants';
export * from './create-audio-item';
export * from './transcripts';
