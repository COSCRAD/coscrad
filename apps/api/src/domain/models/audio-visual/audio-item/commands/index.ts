export * from '../../shared/commands/transcripts';
export * from './constants';
export * from './create-audio-item';
export * from './translate-audio-item-name';
