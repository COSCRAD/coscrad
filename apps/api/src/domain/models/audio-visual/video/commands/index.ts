export * from './create-video';
export * from './translate-video-name';
