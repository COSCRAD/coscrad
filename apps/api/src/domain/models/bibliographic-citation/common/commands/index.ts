export * from './base-create-bibliographic-citation.command-handler';
