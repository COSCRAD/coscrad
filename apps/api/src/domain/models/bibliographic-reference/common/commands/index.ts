export * from './base-create-bibliographic-reference.command-handler';
