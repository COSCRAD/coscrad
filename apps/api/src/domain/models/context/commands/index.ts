export * from './add-audio-for-note';
export * from './connect-resources-with-note';
export * from './create-note-about-resource';
export * from './translate-note';
