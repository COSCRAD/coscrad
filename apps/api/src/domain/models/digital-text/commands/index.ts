export * from './add-audio-for-digital-text-page';
export * from './add-page-to-digital-text/add-page-to-digital-text.command';
export * from './add-page-to-digital-text/add-page-to-digital-text.command-handler';
export * from './add-page-to-digital-text/page-added-to-digital-text.event';
export * from './create-digital-text.command';
export * from './create-digital-text.command-handler';
export * from './digital-text-created.event';
export * from './translate-digital-text-page-content';
