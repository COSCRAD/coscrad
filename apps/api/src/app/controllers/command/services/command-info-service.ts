import { ICommandFormAndLabels } from '@coscrad/api-interfaces';
import { CommandHandlerService, CommandMetadataBase } from '@coscrad/commands';
import { ClassSchema, getCoscradDataSchema } from '@coscrad/data-types';
import { Injectable } from '@nestjs/common';
import { Aggregate } from '../../../../domain/models/aggregate.entity';
import { isNullOrUndefined } from '../../../../domain/utilities/validation/is-null-or-undefined';
import { DomainModelCtor } from '../../../../lib/types/DomainModelCtor';
import { buildCommandForm } from '../../../../view-models/dynamicForms';
import { INDEX_SCOPED_COMMANDS } from '../command-info/constants';

type CommandTypeFilter = (commandType: string) => boolean;

const buildCommandTypeFilter = (
    context?: DomainModelCtor | DetailScopedCommandWriteContext
): CommandTypeFilter => {
    if (isNullOrUndefined(context)) return (_: string) => true;

    if (isCommandWriteContext(context))
        return (commandType: string) => context.getAvailableCommands().includes(commandType);

    return (commandType: string) =>
        // DO NOT DEFAULT TO [] here. Failing to decorate the Resource class should break things!
        Reflect.getMetadata(INDEX_SCOPED_COMMANDS, context).includes(commandType);
};

type CommandInfo = CommandMetadataBase & {
    label: string;
    description: string;
    schema: ClassSchema;
};

export type DetailScopedCommandWriteContext = Aggregate;

type IndexScopedCommandWriteContext = DomainModelCtor;

export const isCommandWriteContext = (input: unknown): input is DetailScopedCommandWriteContext =>
    typeof (input as DetailScopedCommandWriteContext).getAvailableCommands === 'function';

export type CommandContext = IndexScopedCommandWriteContext | DetailScopedCommandWriteContext;

@Injectable()
export class CommandInfoService {
    constructor(private readonly commandHandlerService: CommandHandlerService) {}

    getCommandSchemasWithMetadata(): CommandInfo[] {
        return this.commandHandlerService
            .getAllCommandCtorsAndMetadata()
            .map(({ constructor: ctor, meta }) => ({
                ...meta,
                schema: getCoscradDataSchema(ctor),
            })) as CommandInfo[];
    }

    // TODO Unit test the filtering logic
    getCommandForms(): ICommandFormAndLabels[];
    getCommandForms(context: DomainModelCtor): ICommandFormAndLabels[];
    getCommandForms(context: DetailScopedCommandWriteContext): ICommandFormAndLabels[];
    getCommandForms(context: CommandContext): ICommandFormAndLabels[];
    getCommandForms(context?: CommandContext): ICommandFormAndLabels[] {
        const commandTypeFilter = buildCommandTypeFilter(context);

        return this.getCommandSchemasWithMetadata()
            .filter(({ type }) => commandTypeFilter(type))
            .map(({ label, description, schema, type }) => ({
                label,
                description,
                type,
                form: buildCommandForm(schema, context),
            }));
    }
}
