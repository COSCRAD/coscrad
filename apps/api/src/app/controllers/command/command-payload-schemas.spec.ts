import { CommandModule } from '@coscrad/commands';
import generateDatabaseNameForTestSuite from '../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import createTestModule from '../__tests__/createTestModule';
import { CommandInfoService } from './services/command-info-service';

describe('command payload schemas', () => {
    let commandInfoService: CommandInfoService;

    beforeAll(async () => {
        const testModule = await createTestModule({
            ARANGO_DB_NAME: generateDatabaseNameForTestSuite(),
        });

        await testModule.get<CommandModule>(CommandModule).onApplicationBootstrap();

        commandInfoService = testModule.get<CommandInfoService>(CommandInfoService);
    });

    describe(`Command payload schema`, () => {
        it(`should match the snapshot`, () => {
            const schema = commandInfoService.getCommandSchemasWithMetadata();

            expect(schema).toMatchSnapshot();
        });
    });
});
