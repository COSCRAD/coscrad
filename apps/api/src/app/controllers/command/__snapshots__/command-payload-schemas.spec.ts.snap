// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command payload schemas The schema for command ADD_SONG should have the expected value 1`] = `
Object {
  "audioURL": Object {
    "isOptional": false,
    "type": "URL",
  },
  "contributorAndRoles": Object {
    "isOptional": false,
    "type": Object {
      "contributorId": Object {
        "isOptional": false,
        "type": "NON_EMPTY_STRING",
      },
      "role": Object {
        "isOptional": false,
        "type": "NON_EMPTY_STRING",
      },
    },
  },
  "id": Object {
    "isOptional": false,
    "type": "UUID",
  },
  "lengthMilliseconds": Object {
    "isOptional": false,
    "type": "NonNegativeFiniteNumber",
  },
  "lyrics": Object {
    "isOptional": false,
    "type": "NON_EMPTY_STRING",
  },
  "title": Object {
    "isOptional": true,
    "type": "NON_EMPTY_STRING",
  },
  "titleEnglish": Object {
    "isOptional": true,
    "type": "NON_EMPTY_STRING",
  },
}
`;
